mmv1/third_party/terraform/services/artifactregistry/data_source_artifact_registry_docker_image.go

```diff
@@ -0,0 +1,331 @@
+						if iterTag == tag {
+							res = image
+							break out
+						}
+					}
+				}
+			} else { // use the first image in the response
+				if len(resFiltered) > 0 {
+					res = resFiltered[0]
+					break out
+				}
+			}
+
+			// No matching image was found on this page and no more pages are available
+			if token == "" {
+				return fmt.Errorf("Requested image was not found.")
+			}
+
+			// Retrieve the next page and search again
+			urlNext, err := transport_tpg.AddQueryParams(urlRequest, map[string]string{"pageToken": token})
+			if err != nil {
+				return err
+			}
+
+			resListImages, token, err = retreiveListOfDockerImages(config, project, urlNext, userAgent)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// set the schema data using the response
+	if err := d.Set("name", res.name); err != nil {
+		return fmt.Errorf("Error setting name: %s", err)
+	}
+
+	if err := d.Set("self_link", res.uri); err != nil {
+		return fmt.Errorf("Error setting self_link: %s", err)
+	}
+
+	if err := d.Set("tags", res.tags); err != nil {
+		return fmt.Errorf("Error setting tags: %s", err)
+	}
+
+	if err := d.Set("image_size_bytes", res.imageSizeBytes); err != nil {
+		return fmt.Errorf("Error setting image_size_bytes: %s", err)
+	}
+
+	if err := d.Set("media_type", res.mediaType); err != nil {
+		return fmt.Errorf("Error setting media_type: %s", err)
+	}
+
+	if err := d.Set("upload_time", res.uploadTime); err != nil {
+		return fmt.Errorf("Error setting upload_time: %s", err)
+	}
+
+	if err := d.Set("build_time", res.buildTime); err != nil {
+		return fmt.Errorf("Error setting build_time: %s", err)
+	}
+
+	if err := d.Set("update_time", res.updateTime); err != nil {
+		return fmt.Errorf("Error setting update_time: %s", err)
+	}
+
+	id := fmt.Sprintf("projects/%s/locations/%s/repositories/dockerImages/%s", project, d.Get("region"), d.Get("image"))
+	d.SetId(id)
+
+	return nil
+}
+
+func retreiveListOfDockerImages(config *transport_tpg.Config, project string, urlRequest string, userAgent string) ([]DockerImage, string, error) {
+	resList, err := transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
+		Config:    config,
+		Method:    "GET",
+		Project:   project,
+		RawURL:    urlRequest,
+		UserAgent: userAgent,
+	})
+	if err != nil {
+		return make([]DockerImage, 0), "", err
+	}
+
+	if nextPageToken, ok := resList["nextPageToken"].(string); ok {
+		return flattenDatasourceListResponse(resList), nextPageToken, nil
+	} else {
+		return flattenDatasourceListResponse(resList), "", nil
+	}
+}
+
+func flattenDatasourceListResponse(res map[string]interface{}) []DockerImage {
+	var dockerImages []DockerImage
+
+	resDockerImages, _ := res["dockerImages"].([]interface{})
+
+	for _, resImage := range resDockerImages {
+		image, _ := resImage.(map[string]interface{})
+		dockerImages = append(dockerImages, convertResponseToStruct(image))
+	}
+
+	return dockerImages
+}
+
+func convertResponseToStruct(res map[string]interface{}) DockerImage {
```

Comment:
On first read, I'm wondering if this function is needed. If you didn't have it, and tweak the `d.Set(...)` calls, that might be slightly more straight-forward / consistent with other resources.

Reply:
I think you could do a `d.Set()` call instead of the `dockerImage.tags = stringTags` line you have, but keep the rest.

Reply:
I'm not sure I follow what you're suggesting?  Isn't `d.Set()` specific to the `*schema.ResourceData` type? Here I'm setting fields in the `DockerImage` struct.